### PR TITLE
Complete insertAtCaret refactor

### PR DESCRIPTION
**Pull request by nasdan**

### CODE REVIEW

File: 02 Toolbar/src/common/managers/textAreaManager.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+
+// Based on: http://stackoverflow.com/questions/1064089/inserting-a-text-where-cursor-is-using-javascript-jquery
+class TextAreaManager {
```
Comment (brauliodiez) [inline 02 Toolbar/src/common/managers/textAreaManager.ts]: I don't like neither textAreaUtil and Manager

Let's think for a better name, proposal:

_src/common/ui/tools/textArea.ts_

and the file name

textAreaTools ? 

Maybe Tools or Utils is too generic, but "Manager" sounds weird.

textAreaInsert

File: 02 Toolbar/src/editor.tsx [hunk 2]
```diff
@@ -1,104 +1,80 @@
...
+  offset: number;
+  cursorPosition: number;
 
   constructor(props: Props) {
     super(props);
 
-    this.state = {content: 'test'};
+    this.state = {
+      editorContent: 'test',
+      viewerContent: 'test',
```
Comment (brauliodiez) [inline 02 Toolbar/src/editor.tsx]: Add some simple markdown content for the sample (just to get something already showing)

File: 02 Toolbar/src/editor.tsx [hunk 3]
```diff
@@ -1,104 +1,80 @@
...
-  			var range = document['selection'].createRange();
-  			range.moveStart ('character', -txtarea['value'].length);
-  			strPos = range.text.length;
-  		} else if (br == "ff") {
-  			strPos = txtarea['selectionStart'];
-  		}
+  componentDidUpdate() {
+    if (this.state.shouldUpdateCursor) {
+      textAreaManager.placeCursor(this.textArea, this.cursorPosition);
+    }
```
Comment (brauliodiez) [inline 02 Toolbar/src/editor.tsx:41]: Why is this needed ? Is there any edge case or something to take into account?

File: 02 Toolbar/src/editor.tsx (hunk 2 above)
Comment (nasdan) [reply 02 Toolbar/src/editor.tsx]: Ok

File: 02 Toolbar/src/editor.tsx (hunk 3 above)
Comment (nasdan) [reply 02 Toolbar/src/editor.tsx:41]: Because when react update state and then component, it places cursor to the end of text

File: 02 Toolbar/src/common/managers/textAreaManager.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  insertAtCaret(textArea: HTMLTextAreaElement, caret: string, offsetCursor: number = 0): string {
```
Comment (brauliodiez) [inline 02 Toolbar/src/common/managers/textAreaManager.ts]: We have to think for a better name for this method, now it doesn't update the html element text any more.

Any idea?

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 1 above)
Comment (nasdan) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: src/common/ui/business/textAreaUIBusiness.ts?

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 4 above)
Comment (nasdan) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: Inside this method I'm using the private `buildTextWithCaretBetweenSelectedText` but I think it is too large. Maybe `buildTextWithCaret`?

File: 02 Toolbar/src/editor.tsx [hunk 5]
```diff
@@ -1,104 +1,80 @@
...
+        viewerContent: MTRC(event.target.value).tree,
+        shouldUpdateCursor: false
       });
   }
 
   render() {
     return (
       <div>
-        <input type="submit" value="Italic" className="btn btn-default" onClick={this.onBoldText.bind(this)} />
+        <input type="submit" value="Italic" className="btn btn-default" onClick={this.onItalicText.bind(this)} />
```
Comment (brauliodiez) [inline 02 Toolbar/src/editor.tsx:71]: Let's create a toolbar component

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 1 above)
Comment (brauliodiez) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: I don't like it, is not business, it's an utility or helper for the text area.

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 4 above)
Comment (brauliodiez) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: Maybe we have to keep it like that _insertAtCaret_

Another option could be _insertAtCaretGetText_ but not sure if makes sense

File: 02 Toolbar/src/editor.tsx (hunk 5 above)
Comment (nasdan) [reply 02 Toolbar/src/editor.tsx:71]: Yes, it's the other branch

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 1 above)
Comment (brauliodiez) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: let's call the folder:

./src/common/ui/utils/textArea.ts

and the class name

TextAreaUtil

I know it's not ideal but rather let's avoid blocking this pull because of this.

Do you agree?

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 4 above)
Comment (brauliodiez) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: Let's go for "insertAtCaretGetText"

File: 02 Toolbar/src/editor.tsx (hunk 5 above)
Comment (brauliodiez) [reply 02 Toolbar/src/editor.tsx:71]: ok, sorry :)

**Comment by brauliodiez** [comment]
Let's introduce the updates and move on, give me a buzz when they are ready (ideally we should think for more options but right now, I think they are good enough to move forward), give me a buzz when this is ready for a second review.

File: 02 Toolbar/src/common/managers/textAreaManager.ts (hunk 1 above)
Comment (brauliodiez) [reply 02 Toolbar/src/common/managers/textAreaManager.ts]: Tool is fine as well, your name is right, let's move on

**Comment by nasdan** [comment]
@brauliodiez  It's ready for a second review
